dali/pipeline/operators/decoder/nvjpeg_decoder_decoupled_api.h

```diff
@@ -0,0 +1,364 @@
+    NVJPEG_CALL(nvjpegDestroy(handle_));
+  }
+
+  using dali::OperatorBase::Run;
+  void Run(MixedWorkspace *ws) override {
+    ParseImagesInfo(ws);
+    ProcessImages(ws);
+  }
+
+ protected:
+  void ParseImagesInfo(MixedWorkspace *ws) {
+    // Parsing and preparing metadata
+    for (int i = 0; i < batch_size_; i++) {
+      const auto &in = ws->Input<CPUBackend>(0, i);
+      const auto *input_data = in.data<uint8_t>();
+      const auto in_size = in.size();
+      const auto file_name = in.GetSourceInfo();
+
+      EncodedImageInfo info;
+      nvjpegStatus_t ret = nvjpegGetImageInfo(handle_,
+                                     static_cast<const unsigned char*>(input_data), in_size,
+                                     &info.c, &info.subsampling,
+                                     info.widths, info.heights);
+
+      info.nvjpeg_support = ret == NVJPEG_STATUS_SUCCESS;
+      if (!info.nvjpeg_support) {
+        try {
+          const auto image = ImageFactory::CreateImage(
+                              static_cast<const uint8 *>(input_data), in_size);
+          const auto dims = image->GetImageDims();
+          info.heights[0] = std::get<0>(dims);
+          info.widths[0] = std::get<1>(dims);
+          info.nvjpeg_support = false;
+          output_info_[i] = info;
+        } catch (const std::runtime_error &e) {
+          DALI_FAIL(e.what() + "File: " + file_name);
+        }
+      } else {
+        if (ShouldUseHybridHuffman(info)) {
+          image_decoders_[i] = decoder_huff_hybrid_;
+          image_states_[i] = decoder_huff_hybrid_state_[i];
+        } else {
+          image_decoders_[i] = decoder_huff_host_;
+          image_states_[i] = decoder_host_state_[i];
+        }
+
+        /*
+        // TODO(spanev): add this function when integrating Crop
+        nvjpegnvjpegDecodeParamsSetROI(decode_params_[pos], offset_x, offset_y, roi_w, roi_h);
+        */
+      }
+      const auto c = static_cast<Index>(NumberOfChannels(output_image_type_));
+      output_shape_[i] = Dims({info.heights[0], info.widths[0], c});
+      output_info_[i] = info;
+    }
+  }
+
+  void ProcessImages(MixedWorkspace* ws) {
+    // Creating output shape and setting the order of images so the largest are processed first
+    // (for load balancing)
+    std::vector<std::pair<size_t, size_t>> image_order(batch_size_);
+    for (int i = 0; i < batch_size_; i++) {
+      image_order[i] = std::make_pair(volume(output_shape_[i]), i);
+    }
+    std::sort(image_order.begin(), image_order.end(),
+              std::greater<std::pair<size_t, size_t>>());
+
+    auto& output = ws->Output<GPUBackend>(0);
+    TypeInfo type = TypeInfo::Create<uint8_t>();
+    output.set_type(type);
+    output.Resize(output_shape_);
+    output.SetLayout(DALI_NHWC);
+
+    CUDA_CALL(cudaEventRecord(master_event_, ws->stream()));
+    for (int i = 0; i < num_threads_; ++i) {
+      CUDA_CALL(cudaStreamWaitEvent(streams_[i], master_event_, 0));
```

Comment:
Do we wait on the same `master_event_` which was recorded two lines above but for different streams?

Reply:
Maybe:
```
CUDA_CALL(cudaStreamWaitEvent(ws->stream(), events_[i], 0));
```

Reply:
Event stream in stream_ wait for all work recorded in ws->stream() to be finished

Reply:
Understood